plugins/parcelport/verbs/readers_writers_mutex.hpp

```diff
@@ -0,0 +1,264 @@
+ *
+ * Readers or writers entering the system after the write lock is queued block,
+ * and the next ticket holder (reader or writer) will unblock when the writer
+ * unlocks. An example, continuing from the last line of the above example:
+ *
+ *                                              writers readers next
+ *      W: ticket 3, writers match      OK      3       3       4
+ *      R: ticket 4, readers no match   block   3       3       5
+ *      R: ticket 5, readers no match   block   3       3       6
+ *      W: ticket 6, writers no match   block   3       3       7
+ *      W: ticket 3, unlock                     4       4       7
+ *      R: ticket 4, readers match      OK      4       5       7
+ *      R: ticket 5, readers match      OK      4       6       7
+ *
+ * The 'next' field is a 2-byte value so the available ticket number wraps at
+ * 64K requests. If a thread's lock request is not granted until the 'next'
+ * field cycles and the same ticket is taken by another thread, we could grant
+ * a lock to two separate threads at the same time, and bad things happen: two
+ * writer threads or a reader thread and a writer thread would run in parallel,
+ * and lock waiters could be skipped if the unlocks race. This is unlikely, it
+ * only happens if a lock request is blocked by 64K other requests. The fix is
+ * to grow the lock structure fields, but the largest atomic instruction we have
+ * is 8 bytes, the structure has no room to grow.
+ */
+// from http://locklessinc.com/articles/locks/
+
+#define atomic_xadd(P, V) __sync_fetch_and_add((P), (V))
+#define cmpxchg(P, O, N)  __sync_bool_compare_and_swap((P), (O), (N))
+#define atomic_inc(P)     __sync_add_and_fetch((P), 1)
+
+// Compiler instruction reordering rwl_barrier_
+#define rwl_barrier_() asm volatile("": : :"memory")
+
+#pragma GCC push_options
+#pragma GCC optimize ("O0")
+
+namespace hpx {
+namespace lcos {
+namespace local {
+
+    class readers_writer_mutex
+    {
+    private:
+
+        typedef union {
+            uint64_t u;
+            struct {
+                uint32_t wr;        // Writers and readers
+            } i;
+            struct {
+                uint16_t writers;   // Now serving for writers
+                uint16_t readers;   // Now serving for readers
+                uint16_t next;      // Next available ticket number
+                uint16_t notused;   // Padding to 8 bytes
+            } s;
+        } readwrite_ticket;
+
+        readwrite_ticket ticket;
+
+    public:
+        readers_writer_mutex() : ticket{0} {}
+
+        //
+        // acquire lock for a unique writer
+        //
+        void lock()
+        {
+            // memory ordering barrier
+            rwl_barrier_();
+
+            uint16_t val = atomic_xadd(&ticket.s.next, 1);
+            while (val != ticket.s.writers) {
+// std::cout << "rw lock sleep write " << val << " " << ticket.s.next << " \n";
+                hpx::util::detail::yield_k(4, nullptr);
```

Comment:
The readers writers lock is used in some critical code where latency could be an issue. I wanted a very fast check and recheck, because this lock should not be held for long periods, so I didn't want a possible suspend, just a true spin wait on the lock. To enable the 'exponential backoff' do I just replace the 4 with a larger number (32 is a key one if I recall correctly), or should I use a different construct?